### PR TITLE
Profiler: annotate NCCL batch_isend_irecv

### LINKED ISSUE #194693: Profiler: add batch-level `record_param_comms` for NCCL `batch_isend_irecv`
## Summary

`torch.distributed.batch_isend_irecv` with the NCCL backend does not emit a
batch-level `record_param_comms` annotation that spans the `ncclGroupEnd()`
which launches the batched P2P kernel.

`ProcessGroupNCCL::send` and `recv` each emit `record_param_comms`, but their
scopes only cover the individual `ncclSend` / `ncclRecv` registration calls.
The actual device work for `batch_isend_irecv` is submitted by `ncclGroupEnd()`
in `ProcessGroupNCCL::endCoalescing()`, after those scopes have ended.

As a result, a profiler trace cannot reliably associate the batched P2P kernel
with a single communication annotation containing process-group metadata.

## Reproduction

Run the following with `torchrun --nproc_per_node=2 repro.py`:

```python
import torch
import torch.distributed as dist


dist.init_process_group("nccl")
rank = dist.get_rank()
torch.cuda.set_device(rank)

send = torch.ones(1024, device="cuda") * rank
recv = torch.empty_like(send)
world_size = dist.get_world_size()

ops = [
    dist.P2POp(dist.isend, send, (rank + 1) % world_size),
    dist.P2POp(dist.irecv, recv, (rank - 1) % world_size),
]

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
) as prof:
    works = dist.batch_isend_irecv(ops)
    for work in works:
        work.wait()
    torch.cuda.synchronize()

prof.export_chrome_trace(f"/tmp/trace-rank{rank}.json")
```

## Actual behavior

The trace contains per-operation `record_param_comms` entries for `send` and
`recv`, but no `record_param_comms` entry for `batch_isend_irecv` whose scope
covers the batched NCCL launch. Consequently, the batch kernel cannot be
reliably attributed to a single annotation with Process Group Name /
Description and aggregate P2P metadata.

## Expected behavior

Emit one batch-level `record_param_comms` annotation in
`ProcessGroupNCCL::endCoalescing()` for P2P-only coalescing regions. Its scope
should cover `ncclGroupEnd()` and include:

- Process-group UID and description.
- `Collective name = "batch_isend_irecv"`.
- P2P sequence number and `async=true`.
- Aggregate communication size and dtype metadata.

This would allow profiler consumers to correlate the device kernel with its
process-group metadata.

## Environment

main

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

### PR DESCRIPTION
**Pull request by zupengwang**
## Issue

Fixes #194693

## Summary

> Add a batch-level `record_param_comms` scope for P2P-only NCCL coalescing so the scope includes `ncclGroupEnd()`. Profiler metadata is accumulated only while callbacks are active, and mixed-dtype batches report `UNKNOWN_SCALAR` rather than an incorrect aggregate dtype.

I reviewed the exact three-file diff at commit `759750cad66a6ab14e389e32d8193173096e8444`. I understand that it addresses the missing batch-to-kernel profiler correlation described in the issue, and I take responsibility for the implementation and tests.

## Test plan

> - Exact-base/patch profiler A/B on 2 x RTX 3090 with CUDA 13.0 and NCCL 2.31.2: baseline had no batch event or correlated kernel; the patch produced one of each per rank with the expected aggregate metadata.
> - Mixed-dtype two-rank reproduction: one batch event and correlated kernel per rank, split sizes `[1024, 256]`, and `UNKNOWN_SCALAR` dtype.
> - `python test/distributed/test_c10d_nccl.py -k test_profiler_nccl_batch_isend_irecv_annotations`
> - `python test/distributed/test_c10d_nccl.py -k test_profiler_nccl_annotations_on_gpu_kernels`
> - `cmake --build build --target torch_python -j32`
> - `clang-format 19.1.4 --dry-run --Werror` on the changed C++ files, `git diff --check`, and Python bytecode compilation.

I reviewed these results. This is a profiler correctness/integration change, not a throughput claim; NCCL2, ROCm, multi-node, and nonblocking-group-end configurations were not exercised locally.

## Checklist

- [ ] Passes lint (`spin fixlint`) — full `spin fixlint` was not run; focused formatting and diff checks passed
- [x] Added/updated tests
- [x] Updated documentation (not applicable)
- [x] Included benchmark results (not applicable; no performance claim)

## BC-breaking?

No.

## AI-assisted development

> Codex assisted with source inspection, implementation, test orchestration, and drafting the quoted portions of this description.

I carefully reviewed the code and validation evidence before authorizing this Draft PR.

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -3709,6 +3751,14 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::endCoalescing(OpType optype) {
   coalescing_state_ = 0;
   coalescedComm_ = nullptr;
   coalescedTensors_.clear();
```
Comment (d4l3k) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:3711]: is this something we can add to TensorShelf instead? worried about holding on to tensors for longer than necessary causing memory spikes

**Review by d4l3k — COMMENTED** [review]
Is this also an issue for nccl2 backend? We just switched the default over so wondering if we should focus fixes on nccl2 instead rather than legacy backend

**Comment by zupengwang** [comment]
I reviewed the updated patch and the local validation results.
Thanks for the pointers. I checked the nccl2 backend. It already keeps one `TracingGuard` around the complete grouped send/recv, so the scope bug from the legacy backend is not present there. The remaining issue is the public profiler label: nccl2 reports the internal name `batch_op_issue` instead of `batch_isend_irecv`.
I reworked the PR to target nccl2: rename the tracing and Work labels to `batch_isend_irecv`, add a 2-rank CPU+CUDA profiler regression test, and remove the legacy-backend changes. That also removes the retained tensor vector, so this version does not extend tensor lifetimes.
On 2 x RTX 3090 with `TORCH_DIST_USE_NCCL2=1`, the baseline produced `batch_op_issue`; the patched build produced exactly one `record_param_comms` event and one CUDA kernel per rank labeled `batch_isend_irecv`. The new test and the existing per-group sequence-number test both pass.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (zupengwang) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:3711]: Addressed by dropping the legacy backend changes entirely and targeting nccl2 instead. The tensor-retaining vectors are no longer part of this PR, so tensor lifetimes are not extended.

**Comment by zupengwang** [comment]
I reviewed the current head (`3767bd08713bd844e24bcb1e0454ffe1b61a6df7`) and confirmed that the final two-file diff directly addresses the nccl2 scope and tensor-lifetime concerns raised in the review.

> **AI-assisted summary:**
>
> @d4l3k, thanks for the review. The current head now targets nccl2 only: it changes the public `TracingGuard` and Work labels from `batch_op_issue` to `batch_isend_irecv`, adds a two-rank profiler regression test, and leaves the legacy backend untouched. The tensor-retaining vectors introduced by the earlier version are absent from the final diff, so this patch does not extend tensor lifetimes relative to `main`.
>
> Existing validation on 2 x RTX 3090 with `TORCH_DIST_USE_NCCL2=1` observed `batch_op_issue` on the baseline and one `record_param_comms` event plus one correlated CUDA kernel per rank labeled `batch_isend_irecv` after the patch. The new test and the existing per-group sequence-number test passed.
>
> Would you be willing to re-review this version?

This summary is relevant because it explains how the current head addresses both points from your review. Full CI has not run yet: the available Actions runs are `action_required`, and no `ciflow/*` label is currently applied.

File: torch/csrc/distributed/c10d/nccl2/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -856,7 +856,7 @@ c10::intrusive_ptr<WorkNCCL> ProcessGroupNCCL::batch_op_issue(
   TracingGuard tracingGuard(
       name_,
       comm_size_,
-      "batch_op_issue",
+      "batch_isend_irecv",
```
Comment (d4l3k) [inline torch/csrc/distributed/c10d/nccl2/ProcessGroupNCCL.cpp:859]: I'm not sure we want to rename this since this is for all coalescing ops not just send/recv

**Review by d4l3k — CHANGES_REQUESTED** [review]
@zupengwang it seems like this actually is already solved with nccl2 then?

File: torch/csrc/distributed/c10d/nccl2/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (zupengwang) [reply torch/csrc/distributed/c10d/nccl2/ProcessGroupNCCL.cpp:859]: You're right — nccl2 already has a batch-level TracingGuard spanning the grouped operation, so the original profiling gap is addressed there. Renaming the generic batch_op_issue label to batch_isend_irecv would be too narrow for the coalescing scope. I'll close this PR. Thanks for the clarification.
